### PR TITLE
[None][feat] Enable KVCacheManagerV2 by default for Gemma3 and Gemma4

### PR DESCRIPTION
**Pull request by erictsai-nv**
## Dev Engineer Review

- Gemma3, Gemma3 VLM, Gemma4, and Gemma4 multimodal models enable `KVCacheManagerV2` by default.
- Gemma4 retains the `FLASHINFER` backend default.
- Gemma3 and Gemma4 select the `"PYTHON"` transceiver runtime for disaggregated NIXL serving.
- The model-level defaults use consistent typed APIs.
- No configuration or test-list files changed.
- No correctness, scope, API consistency, or regression issues were identified.

## QA Engineer Review

- Added `test_gemma3_defaults_resolve_to_v2`.
- Added `test_gemma3_explicit_user_setting_wins`.
- Added `test_gemma4_defaults_resolve_to_v2`.
- Added `test_gemma4_explicit_user_setting_wins`.
- Tests cover Gemma3, Gemma3 VLM, Gemma4 causal, Gemma4 conditional-generation, and Gemma4 multimodal default resolution.
- Tests cover KV cache manager selection, backend selection, transceiver runtime selection, and explicit user-setting precedence.
- No corresponding entries were added to `test-db/` or `qa/`.
- Verdict: sufficient.

## Description

Enables `KVCacheManagerV2` by default for Gemma3 and Gemma4.

All four Gemma entry points — `Gemma3ForCausalLM`, `Gemma3VLM` (`Gemma3ForConditionalGeneration` checkpoints, i.e. 4B/12B/27B), `Gemma4ForCausalLM`, and `Gemma4MultimodalModelBase` — declare `get_preferred_kv_cache_manager_version() -> "V2"` (the per-model preference hook introduced in #16060). `Gemma3VLM` and the Gemma4 entry points also declare `get_preferred_transceiver_runtime() -> "PYTHON"` so that disaggregated serving over NIXL retains V2 instead of falling back to V1 (`Gemma3ForCausalLM` gained the same hook upstream in #16787). The preference is adopted only when the user leaves `kv_cache_config.use_kv_cache_manager_v2` at `"auto"`; an explicit user value always wins.

Notes:
- Gemma4 hybrid attention was already unconditionally routed to V2 via `_non_hybrid_kv_cache_manager_cls` (`_util.py`); this change extends the default to non-hybrid Gemma4 and all Gemma3 variants.
- `Gemma4UnifiedForConditionalGeneration` (12B) is hybrid attention, so it is routed to V2 unconditionally as well — the preference it inherits through `Gemma4MultimodalModelBase` is a no-op for it. Validated on the real 12B checkpoint regardless: text + image outputs are sane, and the cpp/python V2 backends produce bitwise-identical tokens (B200, greedy).
- Per review discussion with @liji-nv , this PR also adds the missing `get_preferred_transceiver_runtime() -> "PYTHON"` to DeepSeek-V4 and MiniMax-M3 (plus `get_preferred_kv_cache_manager_version() -> "V2"` for M3): both run V2-core cache managers that the C++ transceiver cannot drive, and their disaggregated tests already pin NIXL + PYTHON — the fully-"auto" route now resolves to the same combination. MiniMax-M2 is deliberately left without a preference: its disaggregated serving is unvalidated, and the absence keeps that visible.
- Docs: added a Gemma row to the "Selecting the KV Cache Manager" table introduced by #16942.

Since #14047 landed, `TLLM_KV_CACHE_MANAGER_V2_BACKEND` defaults to `cpp`, so this PR moves Gemma onto the C++ V2 core.

### Validation

H100 (agg + disagg) and B200 (NVFP4 MMMU): Gemma3 1B/27B (27B = VLM entry point) and Gemma4 26B-A4B (multimodal entry point) all passed across v1/v2py/v2cpp arms, accuracy within run noise. Gemma4 unified 12B: real-checkpoint smoke on B200, v2py/v2cpp bitwise identical with sane multimodal outputs.

## Test Coverage

- `test_registered_models_prefer_v2` (extended): the five Gemma architectures and the two MiniMax-M3 architectures join the V2-preference list.
- `test_registered_models_keep_v2_on_nixl` (new, next to the list above): for every model preferring both V2 and the Python transceiver, resolve the fully-"auto" NIXL route in production order (transceiver runtime first, then KV cache manager) and assert V2 + PYTHON survive. MiniMax-M2 (disagg unvalidated) is documented exclusion.
- The explicit-user-value precedence and the disagg demotion arms are already covered by the existing mock-based tests in `TestKvCacheManagerV2AutoResolution`.

Existing Gemma3/4 integration tests that do not set `use_kv_cache_manager_v2` explicitly now run through the `"auto"` resolution path and exercise V2 by default.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 1]
```diff
@@ -561,6 +561,9 @@ def get_model_defaults(cls, llm_args) -> dict:
         """Gemma4-specific defaults — see Gemma4ForCausalLM.get_model_defaults."""
         return {
             "attn_backend": "FLASHINFER",
+            "kv_cache_config": {
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py]: The multimodal classes get the V2 default but not `get_preferred_transceiver_runtime()`. In disagg with NIXL, `_resolve_transceiver_runtime_auto` leaves the runtime at CPP, and `_resolve_kv_cache_manager_v2_auto` (`llm_utils.py:610-616`) then downgrades this default back to V1 with only an INFO log — so Gemma4 MM disagg silently keeps V1 while Gemma4 text gets V2. If MM disagg is out of scope here (`[modeling_gemma4mm.py:968](https://github.com/NVIDIA/TensorRT-LLM/pull/17396/files#diff-c199d40bc9ba6c34e1a2ed072172254df5e14e0d33790fae775615591b4a744fR968)` rejects the mm-disagg env path), say so in a comment; otherwise add the same `-> "PYTHON"` override on `Gemma4MultimodalModelBase` so both paths agree.

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 2]
```diff
@@ -285,6 +285,21 @@ def __init__(
                          hidden_size=model_config.pretrained_config.hidden_size,
                          vocab_size=model_config.pretrained_config.vocab_size)
 
+    @classmethod
+    def get_model_defaults(cls, llm_args) -> dict:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma3.py]: This only covers `Gemma3ForCausalLM`. `AutoModelForCausalLM._resolve_class` keys off `architectures[0]`, so `google/gemma-3-{4b,12b,27b}-it` (`Gemma3ForConditionalGeneration` → `Gemma3VLM`, `modeling_gemma3vl.py:178`) hits neither `get_model_defaults` nor `get_preferred_transceiver_runtime` and stays on V1. That contradicts "all Gemma3 variants" plus the 27B validation in the description. If the VLM path is deliberately excluded (e.g. V2 not validated with the bidirectional image mask), please state that; otherwise add both hooks to `Gemma3VLM`.

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py [hunk 3]
```diff
@@ -684,3 +684,12 @@ def test_gemma3_local_context_mask_multi_image(self) -> None:
             ],
             device=device).bool()
         torch.testing.assert_close(attention_mask, expected_attention_mask)
+
+
+def test_gemma3_model_defaults_select_v2():
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_gemma3.py]: These four new tests just re-assert the dict literal on the line above them — they'll pass whether or not the default is ever honored. The behavior that can actually regress is the resolution path: given `use_kv_cache_manager_v2="auto"` and a Gemma3/Gemma4 config, does `load_config_and_apply_defaults` land on V2, and does an explicit user `False` still win? `apply_model_defaults_to_llm_args` / `_resolve_kv_cache_manager_v2_auto` are importable and need no GPU — one test through them would be worth all four of these.

File: tensorrt_llm/_torch/models/modeling_gemma3.py (hunk 2 above)
Comment (erictsai-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py]: Added both hooks to Gemma3VLM. The VLM path was validated with V2 explicitly enabled (multimodal 27B fp8).

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 1 above)
Comment (erictsai-nv) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py]: Thanks, added the PYTHON override on Gemma4MultimodalModelBase so both paths agree.

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py (hunk 3 above)
Comment (erictsai-nv) [reply tests/unittest/_torch/modeling/test_modeling_gemma3.py]: Replaced with tests that replay the model_loader resolution ordering (defaults merge → transceiver auto → V2 auto) on the NIXL route.

**Comment by erictsai-nv** [comment]
Docs: deferring to #16942, which adds a "Selecting the KV Cache Manager" table — it'll be merged soon; I'll rebase and add a Gemma row there.

No Gemma-V2 enablement ticket exists; keeping [None].

**Comment by yizhang-nv** [comment]
Now we use get preferred kvcm like transceiver. Please refer to #16060 
```
    @classmethod
    def get_preferred_kv_cache_manager_version(
        cls, pretrained_config: object | None = None
    ) -> Literal["V2"]:
        """Prefer KV cache manager V2 for DeepSeek-V4."""
        return "V2"
```

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 4]
```diff
@@ -557,12 +560,30 @@ class Gemma4MultimodalModelBase(MultimodalModelMixin, PreTrainedModel):
...
     @classmethod
-    def get_model_defaults(cls, llm_args) -> dict:
+    def get_model_defaults(cls, llm_args: "TorchLlmArgs") -> dict:
         """Gemma4-specific defaults — see Gemma4ForCausalLM.get_model_defaults."""
         return {
             "attn_backend": "FLASHINFER",
         }
 
+    @classmethod
+    def get_preferred_kv_cache_manager_version(cls, pretrained_config: Any = None) -> Literal["V2"]:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:570]: The description notes `Gemma4UnifiedForConditionalGeneration` "was not separately validated", yet it inherits the V2 preference here and is asserted in `test_registered_models_prefer_v2`. Since it has a distinct encoder-free vision path and already has coverage in `l0_b200.yml`, could you either run it through the same validation matrix or document why the shared text backbone makes the Gemma4 MM result transferable?

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py [hunk 5]
```diff
@@ -684,3 +692,40 @@ def test_gemma3_local_context_mask_multi_image(self) -> None:
...
+    _resolve_transceiver_runtime_auto(llm_args, model_cls)
+    assert _resolve_kv_cache_manager_v2_auto(llm_args, model_cls) is True
+
+    assert llm_args.cache_transceiver_config.transceiver_runtime == "PYTHON"
+    assert llm_args.kv_cache_config.use_kv_cache_manager_v2 is True
+
+
+@pytest.mark.parametrize("user_setting", [False, True])
+@pytest.mark.parametrize("model_cls", [Gemma3ForCausalLM, Gemma3VLM])
+def test_gemma3_explicit_setting_wins(model_cls: type,
```
Comment (zhaoyangwang-nvidia) [inline tests/unittest/_torch/modeling/test_modeling_gemma3.py]: `TestKvCacheManagerV2AutoResolution.test_explicit_value_overrides_model_preference` in `tests/unittest/llmapi/test_llm_args.py` is already parametrized over `model_cls` and asserts exactly this, and `test_registered_models_prefer_v2` (which this PR extends) already covers the `get_registered_model_class` assertion. Could the Gemma classes be added to those parametrizations instead? The only genuinely new coverage is the NIXL transceiver-then-V2 ordering, and that contract is generic rather than Gemma-specific, so it would also be better placed alongside the other resolver tests than duplicated in two model test files.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 4 above)
Comment (erictsai-nv) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:570]: Ran the real checkpoint on B200 anyway: sane text + image outputs, and the cpp/python V2 backends produce bitwise-identical tokens. Description updated.

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py (hunk 5 above)
Comment (erictsai-nv) [reply tests/unittest/_torch/modeling/test_modeling_gemma3.py]: Done — the per-model tests are removed and the NIXL ordering now lives in `test_registered_models_keep_v2_on_nixl`next to `test_registered_models_prefer_v2,` as a model list.
